luda-editor/client/src/pages/graphic_asset_manage_page/cg_viewer/part_picker.rs

```diff
@@ -0,0 +1,474 @@
+        cg_part_name: String,
+        cg_part_variant_name: String,
+    },
+    TurnOffCgPartVariant {
+        cg_part_name: String,
+        cg_part_variant_name: String,
+    },
+}
+
+pub enum InternalEvent {
+    MouseHoverInThumbnail { global_xy: Xy<Px>, name: String },
+    MouseHoverOutThumbnail,
+}
+
+impl Component for PartPicker<'_> {
+    fn render(self, ctx: &RenderCtx) -> RenderDone {
+        const PADDING: Px = px(8.0);
+
+        let Self {
+            wh,
+            cg_file,
+            project_id,
+            screen_cg,
+            on_event,
+        } = self;
+
+        let (mouse_hovering_part_variant, set_mouse_hovering_part_variant) =
+            ctx.state::<Option<MouseHoveringPartVariant>>(|| None);
+
+        let on_internal_event = |internal_event| match internal_event {
+            InternalEvent::MouseHoverInThumbnail { global_xy, name } => {
+                set_mouse_hovering_part_variant
+                    .set(Some(MouseHoveringPartVariant { name, global_xy }))
+            }
+            InternalEvent::MouseHoverOutThumbnail => set_mouse_hovering_part_variant.set(None),
+        };
+
+        ctx.compose(|ctx| {
+            let Some(mouse_hovering_part_variant) = mouse_hovering_part_variant.deref() else {
+                return;
+            };
+            ctx.add(mouse_hovering_part_variant.tooltip());
+        });
+
+        ctx.compose(|ctx| {
+            padding(PADDING, |wh, ctx| {
+                ctx.add(CgPartList {
+                    wh,
+                    cg_file,
+                    project_id,
+                    screen_cg,
+                    on_event,
+                    on_internal_event: &on_internal_event,
+                });
+            })(wh, ctx);
+        });
+
+        ctx.component(
+            simple_rect(wh, Color::TRANSPARENT, 0.px(), Color::TRANSPARENT).attach_event(|event| {
+                if let namui::Event::MouseMove { event: _event } = event {
+                    if mouse_hovering_part_variant.is_none() {
+                        return;
+                    }
+                    on_internal_event(InternalEvent::MouseHoverOutThumbnail);
+                }
+            }),
+        );
+
+        ctx.done()
+    }
+}
+
+#[component]
+struct CgPartList<'a> {
+    wh: Wh<Px>,
+    cg_file: &'a CgFile,
+    project_id: Uuid,
+    screen_cg: &'a ScreenCg,
+    on_event: &'a dyn Fn(Event),
+    on_internal_event: &'a dyn Fn(InternalEvent),
+}
+impl Component for CgPartList<'_> {
+    fn render(self, ctx: &RenderCtx) -> RenderDone {
+        let Self {
+            wh,
+            cg_file,
+            project_id,
+            screen_cg,
+            on_event,
+            on_internal_event,
+        } = self;
+
+        let wh = ctx.track_eq(&wh);
+        let max_thumbnails_per_row = ctx.memo(|| {
+            (wh.width / (THUMBNAIL_WH.width + MIN_THUMBNAIL_CONTAINER_SIDE_PADDING * 2.0)).floor()
+                as usize
+        });
```

Comment:
Actually this code is more cost than just divide and floor. track_eq and memo is not free. Don't use them too often before you face the performance issue.

Reply:
I didn't see this comment because it collapsed.

Reply:
Fixed with 268e669